src/syscheckd/registry/registry.c

```diff
@@ -0,0 +1,501 @@
+
+    if (RegOpenKeyEx(root_key_handle, sub_key, 0, access_rights, &current_key_handle) != ERROR_SUCCESS) {
+        mdebug1(FIM_REG_OPEN, sub_key, arch == ARCH_32BIT ? "[x32]" : "[x64]");
+        return;
+    }
+
+    /* We use the class_name, sub_key_count and the value count */
+    if (RegQueryInfoKey(current_key_handle, NULL, NULL, NULL, &sub_key_count, NULL, NULL, &value_count, NULL, NULL,
+                        NULL, &file_time) != ERROR_SUCCESS) {
+        RegCloseKey(current_key_handle);
+        return;
+    }
+
+    /* Query each sub_key and call open_key */
+    for (i = 0; i < sub_key_count; i++) {
+        char new_full_key[MAX_KEY + 2];
+        char *new_sub_key;
+        TCHAR sub_key_name_b[MAX_KEY_LENGTH + 1];
+        DWORD sub_key_name_s = MAX_KEY_LENGTH;
+
+        if (RegEnumKeyEx(current_key_handle, i, sub_key_name_b, &sub_key_name_s, NULL, NULL, NULL, NULL) != ERROR_SUCCESS) {
+            continue;
+        }
+
+        snprintf(new_full_key, MAX_KEY, "%s\\%s", full_key, sub_key_name_b);
+
+        if (new_sub_key = strchr(new_full_key, '\\'), new_sub_key) {
+            new_sub_key++;
+        }
+
+        /* Open sub_key */
+        fim_open_key(root_key_handle, new_full_key, new_sub_key, arch, mode);
+    }
+
+    // Done scanning sub_keys, trigger an alert on the current key if required.
+    new.type = FIM_TYPE_REGISTRY;
+    new.registry_entry.key = fim_registry_get_key_data(current_key_handle, full_key, configuration);
+    new.registry_entry.value = NULL;
+
+    saved.type = FIM_TYPE_REGISTRY;
+    saved.registry_entry.key = fim_db_get_registry_key(syscheck.database, full_key);
+    saved.registry_entry.value = NULL;
+
+    if (saved.registry_entry.key != NULL) {
+        new.registry_entry.key->id = saved.registry_entry.key->id;
+    }
+
+    if (!fim_check_restrict(full_key, configuration->filerestrict)) {
+        cJSON *json_event = fim_registry_event(&new, &saved, configuration, mode,
+                                               saved.registry_entry.key == NULL ? FIM_ADD : FIM_MODIFICATION, NULL, NULL);
+
+        if (json_event) {
+            if (fim_db_insert_registry_key(syscheck.database, new.registry_entry.key, new.registry_entry.key->id) != FIMDB_OK) {
+                mwarn("Couldn't insert into DB");
+            }
+
+            if (_base_line) {
+                char *json_formated = cJSON_PrintUnformatted(json_event);
+                send_syscheck_msg(json_formated);
+                os_free(json_formated);
+            }
+
+            cJSON_Delete(json_event);
+        }
+
+        fim_db_set_registry_key_scanned(syscheck.database, full_key);
+    }
+
+    if (value_count) {
+        fim_read_values(current_key_handle, &new, &saved, arch, value_count, mode);
+    }
+
+    fim_registry_free_key(new.registry_entry.key);
+    fim_registry_free_key(saved.registry_entry.key);
+    RegCloseKey(current_key_handle);
+    return;
```

Comment:
There is no need for an empty return statement at the end of a void function, please consider removing it.

Reply:
Done